# pituitary_transcriptome_analyses
* The repo contains essential scripts and datasets to reproduce bioinformatics analyses in [Hou &amp; Chan et al. 2022](https://www.biorxiv.org/content/10.1101/2022.01.05.475069v1?rss=1) 

1. Scripts used to process and analyze mouse pituitary UTR-seq data can be found in `UTRseq_analysis/` folder.
   * **Figures 1C, 1D, 2A, 2C, 2D, 2F, 2G, 3B, 4**
   * **Supplementary figures S1C, S2A, S3A, S4, S5**
   * **Table 1, 3**
   * **Supplementary tables S3, S5, S7**
2. Scripts used to run miRNA DE analysis can be found in `miRNA_DE_analysis/` folder.
   * **Figures 1E, 2B, 2E**
   * **Supplementary figures S2B, S3C**
   * **Table 2**
   * **Supplementary tables S2, S4**
3. Scripts used to run miRNA-gene correlation analysis can be found in `miRNA_correlation_analysis/` folder.
   * **Figure 3A**
   * **Supplementary table S6**
4. Scripts used to process LISA results can be found in `LISA_analysis/` folder.
   * **Figure 3C**
5. Scripts used to process and integrate single-nuclei data from Ruf-Zamojski et al. 2021 can be found in `scRNAseq_analysis/` folder.
   * **Supplementary figures S6A, S6B**
5. Scripts used to test for enrichment of co-expression module genes within snRNA-seq data can be found in `scRNAseq_analysis/` folder.
   * **Supplementary figure S6C**
6. Scripts used for RNA-seq deconvolution can be found in `scRNAseq_analysis/` folder.
   * **Figure 5A**
7. Scripts used to identify sex-biased genes in cell types using scMappR can be found in `scRNAseq_analysis/` folder.
   * **Figure 5B**
   * **Supplementary table S8**


