# pituitary_transcriptome_analyses
The repo contains essential scripts and datasets to reproduce bioinformatics analyses in [Hou &amp; Chan et al. 2022](https://www.biorxiv.org/content/10.1101/2022.01.05.475069v1?rss=1) 

1. Scripts used to process and analyze mouse pituitary UTRseq data can be found in `UTRseq_analysis/` folder.
   * **Figures 1C, 1D, 2A, 2C, 2D, 2F, 2G, 3B, 4**
   * **Supplementary figures S1C, S2A, S3A, S4, S5**
   * **Tables 1, 3**
   * **Supplementary table S3, S5, S7**
2. Scripts used to reprocess single-cell data from Cheung et al. 2018 can be found in `scRNAseq_analysis/` folder.
   * *To run this script, download data from [GSE120410](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120410) into `scRNAseq_analysis/input_files/`*
   * **Figures 5A, 5B, 5C**
   * **Supplementary figures S6, S7**
3. Scripts used to run scMappR can be found in `scRNAseq_analysis/` folder.
   * *Run single-cell reprocessing scripts (2.) prior to running this script*
   * **Figure 5D**
   * **Supplementary table S8, S9**
4. Scripts used to run miRNA DE analysis can be found in `miRNA_DE_analysis/` folder.
   * **Figures 1E, 2B, 2E**
   * **Supplementary figures S2B, S3C**
   * **Table 2**
   * **Supplementary table S2, S4**
5. Scripts used to run miRNA-gene correlation analysis can be found in `miRNA_correlation_analysis/` folder.
   * **Figures 3A**
   * **Supplementary table S6**
7. Scripts used to process LISA results can be found in `LISA_analysis/` folder.
   * **Figures 3C**
