---
title: "Running scMappR"
author: "Cadia Chan"
affiliation: "SickKids Research Institute & University of Toronto"
date: January 11, 2022
output:
 html_document:
   code_folding: hide
   toc: true
   toc_float: true
---
# Note: This script runs on R 4.0 (scMappR R version dependency)
scMappR was updated to v1.0.7 (2021-10-16 ver).
Ensure that Seurat version is 3.2.3.

This script runs scMappR (https://cran.r-project.org/web/packages/scMappR/index.html) to identify 
cell-type specific DE genes by calculating cell-weighted fold-changes (cwFC) for bulk DE genes with cell proportions estimated by single cell deconvolution.
scMappR is run with sex-biased genes identified at each age.
cwFC calculated by scMappR is further processed, however, only PD22 and 37 yielded cwFC for DE genes with high confidence.
Genes predicted to be cell-type specific sex-biased genes checked for in 3 gene lists:

1) Sex-biased genes identified in rat anterior pituitary by scRNA-seq (Fletcher et al 2019; https://doi.org/10.3389/fendo.2019.00623)
2) Genes associated with puberty, puberty-related diseases, pituitary diseases.
3) Sex-biased genes identified in the human pituitary by GTEx (Oliva et al 2020; https://doi.org/10.1126/science.aba3066)

# Updates:
* 2022/01/11: Updated cwFoldChange_evaluate_new (old custom function from Dustin) to cwFoldChange_evaluate (official function in scMappR); change made in functions R script.
* 2022/01/12: Added combined heatmap for all ages to show cwFC values
* 2022/01/13: Saved output from FindMarkers and load in as input instead (decreases runtime); setwd for cwFC plots part updated for knitR

To run: Set working directory "To Source File Location".

## Libraries and source scripts

```{r load_library, warning = F, message = F}
#### Libraries ####
library(dplyr)
library(Seurat)
library(reshape2)
library(ggplot2)
library(ggpubr)
library(scMappR)
library(scales)
library(colorspace)
library(ggrepel)
library(openxlsx)
library(biomaRt)
library(pheatmap)
library(knitr)
library(EDASeq)
library(WGCNA)
library(RColorBrewer)

source("scmappr_analysis_functions_2022-01-11_rmd.R")
ensembl_marts <- readRDS("input_files/ensembl_marts.rds")
rat <- ensembl_marts[["rat"]]
mouse <- ensembl_marts[["mouse"]]
```

## Make signature matrix for scMappR
Gene signature matrix is prepared for scMappR using both positive and negative gene markers identified by Seurat FindAllMarkers function.
Signature matrix contains odds ratio calculated by scMappR generes_to_heatmap function.

```{r make_scmappr_OR, warning = F, message = F}
# Make scmappr input (OR matrix) ####
dir.create("output_files/scmappr_20220111/")
pit_seurat_manual <- readRDS("output_files/seurat_v3_cheung_etal_manual_cell_anno.rds")
# 
# man_markers <- FindAllMarkers(pit_seurat_manual, min.pct = 0.25)
# man_markers <- 
#   mutate(man_markers, cluster = gsub(" ", "_", cluster)) %>%
#   mutate(cluster = gsub("/", "_", cluster)) %>%
#   mutate(cluster = gsub("(", "", cluster, fixed = T)) %>%
#   mutate(cluster = gsub(")", "", cluster, fixed = T))
# saveRDS(man_markers, "input_files/seurat_v3_man_markers.rds")
# 
# man_markers$cluster <- factor(man_markers$cluster, levels = c("Somatotropes", "Lactotropes", "Gonadotropes", "Stem-cells_Sox2_FSC",
#                                                               "Proliferating_Pou1f1", "Vascular_Endothelia", "Corticotropes",
#                                                               "Melanotropes_intermediate", "Connective_tissue_VLMC", "Thyrotropes",
#                                                               "Immune_cells", "Pituicytes_posterior"))

man_markers <- readRDS("input_files/seurat_v3_man_markers.rds")
gene_res <- lapply(levels(man_markers$cluster), function(x) man_markers[man_markers$cluster == x, ])
names(gene_res) <- levels(man_markers$cluster)
gene_res <- lapply(gene_res, function(x) {rownames(x) <- as.character(x$gene); x})
gene_res <- lapply(gene_res, function(x) x[, c(5, 2)])

wilcoxon_scmappr <- generes_to_heatmap(gene_res, species = "mouse", make_names = F)
kable(head(wilcoxon_scmappr$OR), caption = "Example gene OR in signature matrix.")
write.table(cbind("GENENAME" = rownames(wilcoxon_scmappr$OR), wilcoxon_scmappr$OR), 
            "output_files/scmappr_20220111/OR_sig_mat.txt",
            quote = F, sep = "\t", col.names = T, row.names = F)
```

## scMappR deconvolution methods comparison
By default, scMappR uses DeconRNASeq as the deconvolution method to estimate cell proportions.
Other deconvolution methods can also be run so that estiamted cell proportions from each method can be compared.
These methods include WGCNA, DCQ and DeconRNASeq.
It is recommended to choose the deconvolution method that best matches prior knowledge of cell proportions.
In our study, we chose DeconRNASeq to estimate cell proportions for downstream analysis as the results are most similar to CIBERSORT estimated cell proportions.

```{r compare_deconvolution, warning = F, message = F, fig.width = 8, fig.height = 6, fig.cap = "Comparison of estimated cell proportions from different deconvolution methods."}
# Compare scMappR deconvolution methods ####
utr_obj <- readRDS("input_files/pit_utr_2019__RUV_k2_set1_2019-07-03.rds")
utr_expr <- normCounts(utr_obj)
decon_method <- compare_deconvolution_methods(utr_expr, signature_matrix = wilcoxon_scmappr$OR)
# print(decon_method)
ggsave("output_files/scmappr_20220111/decon_compare.pdf", last_plot(), device = "pdf",
       width = 8, height = 6)

# gg_def_pal <- hue_pal()(12)
gg_def_pal <- hue_pal()(length(colnames(decon_method$cellType_proportions$DeconRNAseq)))
lighter_cluster_colours <- sapply(gg_def_pal, function(x) lighten(x, amount = 0.5))
cluster_colours <- sapply(gg_def_pal, function(x) darken(x, amount = 0.2))
use_colors <- lapply(1:length(cluster_colours), function(x) c("F" = lighter_cluster_colours[[x]], "M" = cluster_colours[[x]]))
names(use_colors) <- colnames(decon_method$cellType_proportions$DeconRNAseq)

all_decon_plots <- lapply(names(decon_method$cellType_proportions), function(x) make_decon_plots(x,
                                                                                                 file_out="output_files/scmappr_20220111/"))

```
```{r deconrnaseq_plot, warning = F, message = F, fig.width = 14, fig.height = 7, fig.cap = "DeconRNASeq estimated cell proportions."}
print(all_decon_plots[[1]])

```
```{r dcq_plot, warning = F, message = F, fig.width = 14, fig.height = 7, fig.cap = "DCQ estimated cell proportions."}
print(all_decon_plots[[2]])

```
```{r WGCNA_plot, warning = F, message = F, fig.width = 14, fig.height = 7, fig.cap = "WGCNA estimated cell proportions."}
print(all_decon_plots[[3]])
```

## Run scMappR
scMappR is run to calculate cell-weighted fold-change (cwFC) for each DE gene identified by bulk analysis by incorporating estimated cell proportions.
Here, we run scMappR for sex-biased genes from each of the profiled ages (D12, 22, 27, 32, 37).

```{r setup_scmappr, warning = F, message = F}
# Run scMappR ####
bulk_DE_all <- readRDS("input_files/pit_utr_2019_de_result_list_2019-07-03.rds")
bulk_DE <- lapply(bulk_DE_all, function(x) x[abs(x$logFC) > log2(1.5) & x$FDR < 0.05, ])
bulk_normalized <- utr_expr
odds_ratio_in <- wilcoxon_scmappr$OR

meta <- pData(utr_obj)
max_proportion_change <- 10
theSpecies <- "mouse"
# setwd("output_files/")
knitr::opts_knit$set(root.dir = 'output_files/')

```
```{r run_scmappr, warning = F, message = F}
contrast_df <- matrix(c("Pd12F", "Pd12M", "Pd22F", "Pd22M", "Pd27F", "Pd27M", "Pd32F", "Pd32M", "Pd37F", "Pd37M",
                        "Pd22M", "Pd12M", "Pd27M", "Pd22M", "Pd32M", "Pd27M", "Pd37M", "Pd32M",
                        "Pd22F", "Pd12F", "Pd27F", "Pd22F", "Pd32F", "Pd27F", "Pd37F", "Pd32F",
                        "Pd37M", "Pd12M", "Pd37F", "Pd12F", "Pd37M", "Pd22M", "Pd37F", "Pd22F"),
                      nrow = 17, ncol = 2, byrow = T, dimnames = list(names(bulk_DE), c("case", "ctrl")))
contrast_df <- contrast_df[-grep("vs", rownames(contrast_df)), ]

run_scmappr("d37_sex", contrast_df["d37_sex", 1], contrast_df["d37_sex", 2], file_out = "./scmappr_20220111/")
run_scmappr("d32_sex", contrast_df["d32_sex", 1], contrast_df["d32_sex", 2], file_out = "./scmappr_20220111/")
run_scmappr("d27_sex", contrast_df["d27_sex", 1], contrast_df["d27_sex", 2], file_out = "./scmappr_20220111/")
run_scmappr("d22_sex", contrast_df["d22_sex", 1], contrast_df["d22_sex", 2], file_out = "./scmappr_20220111/")
run_scmappr("d12_sex", contrast_df["d12_sex", 1], contrast_df["d12_sex", 2], file_out = "./scmappr_20220111/")
```

## Process cwFC results
scMappR cwFC results are used to select high-confidence genes which are likely cell-type specific DE genes.
These high-confidence DE genes are compared to a list of known cell-type specific sex-biased genes in rat anterior pituitary from Fletcher et al. 2019.
In addition, cell-type specific DE genes which have been implicated in pituitary disease and puberty disease as well as pubertal onset by GWAS are identified.
Finally cell-type specific DE genes which have been reported as sex-biased in human pituitary by GTEx are also identified.
If error is thrown ("cwFoldchange_gene_assigned not found"), then scMappR has found that 'All DEGs are not normally distributed, suggesting very consistent cell-type specificity across genes.'
Hence, no cell-type specific sex-biased genes are identified.

```{r process_cwfc, warning = F, message = F}
# Process the cwFC results ####

names(gg_def_pal) <- c("Somatotropes", "Lactotropes", "Gonadotropes",
                       "Stem-cells_Sox2_FSC", "Proliferating_Pou1f1",
                       "Vascular_Endothelia", "Corticotropes",
                       "Melanotropes_intermediate", "Connective_tissue_VLMC",
                       "Thyrotropes", "Immune_cells", "Pituicytes_posterior")
gg_def_pal_scmappr <- c(gg_def_pal, "not_cw_DE" = "gray50")


fletcher_DE <- read.table("../input_files/2019_Fletcher_rat_ant_pit_sex_bias.txt",sep = "\t", header = T)
# mouse <- useEnsembl(biomart ="genes", dataset = "mmusculus_gene_ensembl")
# rat <- useEnsembl(biomart = "genes", dataset = "rnorvegicus_gene_ensembl")
# saveRDS(list(mouse = mouse, rat = rat), "input_files/ensembl_marts.rds")
check_rno_genes <- getBM(attributes = c("rgd_symbol", "ensembl_gene_id"), filters = "rgd_symbol",
                         values = fletcher_DE$genename, mart = rat)
colnames(check_rno_genes) <- c("genename", "ensembl_gene_id")
check_rno_genes <- full_join(fletcher_DE, check_rno_genes, by = "genename")

pit_pub_genes <- read.table("../input_files/pituitary_puberty_genes_combined_2020-09-28.txt", sep = "\t", header = T)
colnames(pit_pub_genes) <- c("genename", "source", "source_group")

gtex_genes <- melt(readRDS("../input_files/Oliva_science_2021_pit_sex_top500_genes_overlap.rds"),
                   id.vars = c("genename", "ENSEMBL_ID", "HUGO_gene_id", "MASH.beta", "MASH.sd", "MASH.LFSR", "chr"))

```
```{r setscmappr_20220111_dir, message = F}
knitr::opts_knit$set(root.dir = 'output_files/scmappr_20220111/')

```
```{r cwfc_plots_12, message = F, fig.width = 10, fig.cap = "Heatmap of gene-normalized cwFC for hormone-producing cell-type-specific sex-biased genes at PD12."}
# setwd("scmappr_20220111")
cwfc12 <- calc_cwfc_plots(12)
save_pheatmap_pdf(x=cwfc12[1],
                  filename=paste0("scMappR_pituitary_d12", "_sex/cwFC_val_results/d12",
                                  "_cwFC_geneassigned_heatmap_cwFC_20120111.pdf"), width = 10)
write.xlsx(list(cwfc12[2],cwfc12[3], cwfc12[4]), paste0("scMappR_pituitary_d12", "_sex/cwFC_val_results/d12",
                                                        "_cwFC_geneassigned_genelists_overlap_20120111.xlsx"),
           colNames = T,
           sheetName = names(cwfc12[2:4]))


```
```{r cwfc_plots_22, message = F, fig.width = 10, fig.cap = "Heatmap of gene-normalized cwFC for hormone-producing cell-type-specific sex-biased genes at PD22."}
cwfc22 <- calc_cwfc_plots(22)
save_pheatmap_pdf(x=cwfc22[1],
                  filename=paste0("scMappR_pituitary_d22", "_sex/cwFC_val_results/d22",
                                  "_cwFC_geneassigned_heatmap_cwFC_20220111.pdf"), width = 10)
write.xlsx(list(cwfc22[2],cwfc22[3], cwfc22[4]), paste0("scMappR_pituitary_d22", "_sex/cwFC_val_results/d22",
                                                        "_cwFC_geneassigned_genelists_overlap_20220111.xlsx"),
           colNames = T,
           sheetName = names(cwfc22[2:4]))


```
```{r cwfc_plots_27, message = F, fig.width = 10, fig.cap = "Heatmap of gene-normalized cwFC for hormone-producing cell-type-specific sex-biased genes at PD27."}
cwfc27 <- calc_cwfc_plots(27)
save_pheatmap_pdf(x=cwfc27[1],
                  filename=paste0("scMappR_pituitary_d27", "_sex/cwFC_val_results/d27",
                                  "_cwFC_geneassigned_heatmap_cwFC_20220111.pdf"), width = 10)
write.xlsx(list(cwfc27[2],cwfc27[3], cwfc27[4]), paste0("scMappR_pituitary_d27", "_sex/cwFC_val_results/d27",
                                                        "_cwFC_geneassigned_genelists_overlap_20220111.xlsx"),
           colNames = T,
           sheetName = names(cwfc27[2:4]))


```
```{r cwfc_plots_32, message = F, fig.width = 10, fig.cap = "Heatmap of gene-normalized cwFC for hormone-producing cell-type-specific sex-biased genes at PD32."}
cwfc32 <- calc_cwfc_plots(32)
save_pheatmap_pdf(x=cwfc32[1],
                  filename=paste0("scMappR_pituitary_d32", "_sex/cwFC_val_results/d32",
                                  "_cwFC_geneassigned_heatmap_cwFC_20220111.pdf"), width = 10)
write.xlsx(list(cwfc32[2],cwfc32[3], cwfc32[4]), paste0("scMappR_pituitary_d32", "_sex/cwFC_val_results/d32",
                                                        "_cwFC_geneassigned_genelists_overlap_20220111.xlsx"),
           colNames = T,
           sheetName = names(cwfc32[2:4]))


```
```{r cwfc_plots_37, message = F, fig.width = 10, fig.cap = "Heatmap of gene-normalized cwFC for hormone-producing cell-type-specific sex-biased genes at PD37."}
cwfc37 <- calc_cwfc_plots(37)
save_pheatmap_pdf(x=cwfc37[1],
                  filename=paste0("scMappR_pituitary_d37", "_sex/cwFC_val_results/d37",
                                  "_cwFC_geneassigned_heatmap_cwFC_20220111.pdf"), width = 10)
write.xlsx(list(cwfc37[2],cwfc37[3], cwfc37[4]), paste0("scMappR_pituitary_d37", "_sex/cwFC_val_results/d37",
                                                        "_cwFC_geneassigned_genelists_overlap_20220111.xlsx"),
           colNames = T,
           sheetName = names(cwfc37[2:4]))


```
```{r cwfc_heatmap_combined, message = F, fig.width = 20}
cwfc_combined <- list(d12 = cwfc12[[5]], d22 = cwfc22[[5]], d27 = cwfc27[[5]], d32 = cwfc32[[5]], d37 = cwfc37[[5]])
cwfc_combined <- bind_rows(lapply(names(cwfc_combined), function(x) mutate(as.data.frame(cwfc_combined[[x]]), age = x, gene = rownames(cwfc_combined[[x]]))))
cwfc_melt <- melt(cwfc_combined)
colnames(cwfc_melt) <- c("age", "gene", "celltype", "cwFC")

cwfc_melt <- arrange(cwfc_melt, cwFC, celltype)

cwfc_melt_combined <- mutate(cwfc_melt, age_cell = paste0(celltype, "_", age))
geneorder <- unique(cwfc_melt$gene)
agecellorder <- paste0(rep(c("Somatotropes", "Lactotropes", "Gonadotropes", "Corticotropes",
                             "Melanotropes_intermediate", "Thyrotropes"), each=5), "_",
                       rep(c("d12", "d22", "d27", "d32", "d37"), 6))
use_breaks <- get_htmap_breaks(cwfc_melt_combined$cwFC,
                               colorRampPalette(c("orange", "white", "darkmagenta"))(255))
cwfc_melt_combined_df <- dcast(cwfc_melt_combined, gene~age_cell, value.var = "cwFC")
cwfc_melt_combined_df <- cwfc_melt_combined_df[order(match(cwfc_melt_combined_df$gene, geneorder)), ]
rownames(cwfc_melt_combined_df) <- cwfc_melt_combined_df$gene
cwfc_melt_combined_df <- dplyr::select(cwfc_melt_combined_df, -gene)

cwfc_melt_combined_df <- cwfc_melt_combined_df[, order(match(colnames(cwfc_melt_combined_df), agecellorder))]
cwfc_melt_combined_df[is.na(cwfc_melt_combined_df)] <- 0

row_anno <- data.frame(rep(c("Somatotropes", "Lactotropes", "Gonadotropes", "Corticotropes",
                             "Melanotropes_intermediate", "Thyrotropes"), each=5),
                       rep(c("d12", "d22", "d27", "d32", "d37"), 6),
                       paste0(rep(c("Somatotropes", "Lactotropes", "Gonadotropes", "Corticotropes",
                                    "Melanotropes_intermediate", "Thyrotropes"), each=5), "_",
                              rep(c("d12", "d22", "d27", "d32", "d37"), 6)))
colnames(row_anno) <- c("celltype", "age", "cellage")
rownames(row_anno) <- row_anno$cellage
row_anno <- dplyr::select(row_anno, -cellage)
agepal <- brewer.pal(n = 9, "BuGn")[c(1,3,5,7,9)]
names(agepal) <- unique(row_anno$age)
row_anno_pal <- list(age = agepal,
                     celltype = gg_def_pal[names(gg_def_pal) %in% c("Somatotropes", "Lactotropes", "Gonadotropes", "Corticotropes",
                                                                    "Melanotropes_intermediate", "Thyrotropes")])
p <- pheatmap(t(cwfc_melt_combined_df),
              # scale = "column",
              cluster_cols = T,
              cluster_rows = F,
              color = colorRampPalette(c("orange", "white", "darkmagenta"))(255),
              breaks = use_breaks,
              gaps_row = c(5,10,15,20,25,30),
              annotation_row = row_anno,
              annotation_colors = row_anno_pal,
              cellheight = 10, cellwidth = 8)

save_pheatmap_pdf(list(p), filename="cwFC_geneassigned_heatmap_cwFC_combined_20220111.pdf",
                  width = 17, height = 8)
```

## Session Info

```{r session_info}
sessionInfo()
```

